clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -273,26 +273,29 @@ public RecordAppendResult append(String topic,
 
                 // check if we have an in-progress batch
                 Deque<ProducerBatch> dq = topicInfo.batches.computeIfAbsent(effectivePartition, k -> new ArrayDeque<>());
+                RecordAppendResult appendResult;
                 synchronized (dq) {
                     // After taking the lock, validate that the partition hasn't changed and retry.
                     if (topicInfo.builtInPartitioner.isPartitionChanged(partitionInfo)) {
                         log.trace("Partition {} for topic {} switched by a concurrent append, retrying",
                                 partitionInfo.partition(), topic);
                         continue;
                     }
-                    RecordAppendResult appendResult = tryAppend(timestamp, key, value, headers, callbacks, dq, nowMs);
-                    if (appendResult != null) {
+                    appendResult = tryAppend(timestamp, key, value, headers, callbacks, dq, nowMs);
+                    if (appendResult != null && !appendResult.newBatchCreated) {
                         topicInfo.builtInPartitioner.updatePartitionInfo(partitionInfo, appendResult.appendedBytes, cluster);
                         return appendResult;
                     }
                 }
 
-                // we don't have an in-progress record batch try to allocate a new batch
-                if (abortOnNewBatch) {
-                    // Return a result that will cause another call to append.
+                // either 1. current topicPartition producerBatch is full - return and prepare for another batch/partition.
+                // 2. no producerBatch existed for this topicPartition, create a new producerBatch.
+                if (appendResult == null && abortOnNewBatch) {
```

Comment:
It looks like there are still cases when onNewBatch logic is invoked and partition could be switched for the second time, and it seems to change the behavior of DefaultPartitioner - if the batch is sent (removed from the queue) before a new record is produced, onNewBatch isn't going to get called and it'll get stuck with the current partition.

Reply:
@artemlivshits IIUC onNewBatch should only be called [_when existing batch is full_](https://github.com/apache/kafka/blob/b392cf212f7ed4a82b79c3690b488619c027dba9/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java#L79).

The difference this PR is trying to make with this condition [here](https://github.com/apache/kafka/blob/a0e9020f87b0f5014506a32f41bc431c1cbfbc88/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L293-L296) is - 
1. if `appendResult` is null, that means existing batch for current partition is full -> in this case, return and call onNewBatch.
2. if `appendResult` isn't null, at this point will also imply `appendResult.newBatchCreated` [is true](https://github.com/apache/kafka/blob/a0e9020f87b0f5014506a32f41bc431c1cbfbc88/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L285-L288). And producerBatch isn't completed for current partition, onNewbatch should *not* be called. Instead we should stick to this partition, create a producerBatch and fill it. We shouldn't switch partitions here. 

wdyt ?

Reply:
QQ - if there are no batches in current partition, are there cases we would still want to switch partitions? Technically, the partition is completely empty at this point, and we can create and store new batches. I'll revisit the code and get back asap if there exists some case where we wouldn't want to fill empty `batches` in partitions and instead want to switch. 
Apart from that, here is some context on intended change from my limited knowledge on Partitioning logic ->

@artemlivshits IIUC onNewBatch should only be called [when existing batch is full](https://github.com/apache/kafka/blob/b392cf212f7ed4a82b79c3690b488619c027dba9/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java#L79).

The difference this PR is trying to make with this condition [here](https://github.com/apache/kafka/blob/a0e9020f87b0f5014506a32f41bc431c1cbfbc88/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L293-L296) is -

if appendResult is null, that means existing batch for current partition is full -> in this case, return and call onNewBatch.
if appendResult isn't null, at this point will also imply appendResult.newBatchCreated [is true](https://github.com/apache/kafka/blob/a0e9020f87b0f5014506a32f41bc431c1cbfbc88/clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java#L285-L288). And producerBatch isn't completed for current partition, onNewbatch should not be called. Instead we should stick to this partition, create a producerBatch and fill it. We shouldn't switch partitions here.
wdyt ?

Reply:
> if there are no batches in current partition, are there cases we would still want to switch partitions?

It could happen that batch got sent out by the the time next record arrives.  When linger.ms=0 (the default), the sender would try to send the batch out as soon as possible.  So the sequence of events could look like this:

0. batch queue is empty.
1. append record: create a new batch, but don't switch partitions because queue is empty
2. sender drains the queue, so now queue is empty
3. append record: create a new batch, but don't switch partitions because queue is empty
...

Reply:
Makes sense! with current approach we could get stuck to same partition, leading to starvation. 
There are now two constraints - 1. Prevent even partitions getting skipped consistently when there are no producerBatches corresponding to this partition. 2. Do not always select the same partition when producerBatch is empty. ie, detect if recent drain was done for this partition and skip it in that case.

Reply:
with these conditions, i have made some modifications to the original PR, will summarize the new approach.

Reply:
@artemlivshits the approach now is little brittle tbh. We could have had a timeout after drain for a partition, within which partition should get skipped. But that approach seemed to make the solution complex. Wdyt? 
Any other approach you think we should think about instead ?

Reply:
@sudeshwasnik 
I am following up with this fix to make kafka to have proper roundrobin. i had a thought why cant we put the partitionid in record value..so if the partitioner called twice the paritioner in record will be there and twice call of partitioner wont happen.

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1387

if (record.partition() != null)
            return record.partition();
this always returns null.